### PR TITLE
PRO-2076 Send ERC-20

### PR DESCRIPTION
**Pull request by lalitkumarrock**
https://linear.app/pillarproject/issue/PRO-2076/story-task-send-erc-20

## Description

- Made new component Send ERC-20 token( <SendErc20 /> )

## How Has This Been Tested?

- Tested on local

## Screenshots (if appropriate):
Screenshot 2024-01-09 at 8 02 22 PM

https://github.com/etherspot/ui/assets/82652040/b2abb496-60f8-4c2b-aaa3-1c989c5dcd3f

https://github.com/etherspot/ui/assets/82652040/5cab3cca-06e3-48c6-9a41-a96a68569f07

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by lalitkumarrock** [comment]
@IAmKio @bhumitaylor14

File: src/components/SendErc20/Erc20Input.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    <>
+      <input
+        type="text"
+        name="value"
+        id="value"
+        value={value}
+        onChange={onChange}
+        className={className}
+        onKeyDown={onKeyDown}
+        placeholder="0.00"
```
Comment (IAmKio) [inline src/components/SendErc20/Erc20Input.tsx]: Just a note that generally ERC-20 tokens have 18 decimal places and are transferred in multiples of 1. This is better as just `0` and the user can use decimans if they want.

File: src/components/SendErc20/Erc20Input.tsx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  const onKeyDown = () => {
+    if (!disableSendOnEnter) handleEnterPress && handleEnterPress();
+  };
+
+  return (
+    <>
+      <input
+        type="text"
+        name="value"
+        id="value"
```
Comment (IAmKio) [inline src/components/SendErc20/Erc20Input.tsx]: As far as i'm aware all ID's of HTML elements should be unique. Something like `id={`etherspot-ui-erc-20-${_.uniq()}`}` might help

File: src/components/SendErc20/Erc20Input.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        value={value}
+        onChange={onChange}
+        className={className}
+        onKeyDown={onKeyDown}
+        placeholder="0.00"
+      />
+      <p className={`sm:text-sm ${errorMessageClassName}`}>{errorMessage}</p>
```
Comment (IAmKio) [inline src/components/SendErc20/Erc20Input.tsx]: Please remove the `sm:text-sm` class here, i know how you're thinking but they can define `sm:text-sm` via `errorMessageClassName`

File: src/components/SendErc20/EtherspotContract.tsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      {
+        name: '',
+        type: 'bool',
+      },
+    ],
+    payable: false,
+    stateMutability: 'nonpayable',
+    type: 'function',
+  },
+];
```
Comment (IAmKio) [inline src/components/SendErc20/EtherspotContract.tsx]: There may be more of these in future, for example transferring an NFT, the ABI isn't the same. You might want to consider moving this into a folder named `/data/abis` so we can collect future ABI's here and make them sharable via imports.

File: src/components/SendErc20/index.tsx [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        estimatedResponse = await estimate();
+        onEstimated && onEstimated(estimatedResponse);
+      } catch (e) {
+        errorLog('Estimate failed!', e, debug);
+      }
+
+      if (!isEmpty(estimatedResponse) && !onlyEstimate) {
+        /**
+         * The send function simply sends all the EtherspotBatches, 
+           which contain all EtherspotBatch and EtherspotTransaction components to the blockchain via the Etherspot platform.
```
Comment (IAmKio) [inline src/components/SendErc20/index.tsx]: Please wrap this line

File: src/components/SendErc20/index.tsx [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      if (!isEmpty(estimatedResponse) && !onlyEstimate) {
+        /**
+         * The send function simply sends all the EtherspotBatches, 
+           which contain all EtherspotBatch and EtherspotTransaction components to the blockchain via the Etherspot platform. 
+        */
+        try {
+          const response = await send();
+          onSent && onSent(response);
+        } catch (e) {
+          errorLog('Send failed!', e, debug);
```
Comment (IAmKio) [inline src/components/SendErc20/index.tsx]: For all error logs, please ensure that a link to the documentation is also logged:

```suggestion
          errorLog('Etherspot UI: Sorry there was an error whilst sending your transaction. Please visit our documentation for more information: https://etherspot.fyi', e, debug);
```

File: src/components/SendErc20/index.tsx [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        }
+      }
+
+      setIsEstimate(false);
+      onExecutionStatus && onExecutionStatus(false);
+    } catch (e) {
+      setIsEstimate(false);
+      onExecutionStatus && onExecutionStatus(false);
+      errorLog('Send or Estimate failed!', e, debug);
```
Comment (IAmKio) [inline src/components/SendErc20/index.tsx]: As above

File: src/utils/bigNumber.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * This function can be invoke after user inputing consecutive charaters OR user pasting whole string.
+ */
+export function normalizeRawValue(input: string, decimals: number = 18): string {
+  let result = input.replace(/,/g, '.').replace(/\s/g, '');
+
+  const maxDecimals = decimals === 6 ? 4 : decimals > 15 ? 15 : decimals;
+
+  // Trip leading and trailing spaces
```
Comment (IAmKio) [inline src/utils/bigNumber.ts]:
```suggestion
  // Strip leading and trailing spaces
```

File: src/utils/validation.ts [hunk 9]
```diff
@@ -1,9 +1,48 @@
...
   if (!address) return false;
 
   return ethers.utils.isAddress(address);
-
 };
 
+export const erc20ValidationMessage = (props: SendERC20TransactionProps): string => {
+  const { tokenAddress, receiverAddress, value } = props;
+  if (!isValidEthereumAddress(tokenAddress)) {
+    return 'Token addess is not valid!';
```
Comment (IAmKio) [inline src/utils/validation.ts]:
```suggestion
    return 'Token address is not valid!';
```

File: src/utils/validation.ts [hunk 10]
```diff
@@ -1,9 +1,48 @@
...
+  }
+  if (!isValidEthereumAddress(receiverAddress)) {
+    return 'Receiver addess is not valid!';
```
Comment (IAmKio) [inline src/utils/validation.ts]:
```suggestion
    return 'Receiver address is not valid!';
```

**Review by IAmKio — CHANGES_REQUESTED** [review]
A few minor points but overall, great work! 🎉

File: src/components/SendErc20/Erc20Input.tsx (hunk 1 above)
Comment (lalitkumarrock) [reply src/components/SendErc20/Erc20Input.tsx]: Fixed it.

File: src/components/SendErc20/Erc20Input.tsx (hunk 2 above)
Comment (lalitkumarrock) [reply src/components/SendErc20/Erc20Input.tsx]: Updated it.

File: src/components/SendErc20/Erc20Input.tsx (hunk 3 above)
Comment (lalitkumarrock) [reply src/components/SendErc20/Erc20Input.tsx]: Fixed it.

File: src/components/SendErc20/EtherspotContract.tsx (hunk 4 above)
Comment (lalitkumarrock) [reply src/components/SendErc20/EtherspotContract.tsx]: Updated it.

File: src/components/SendErc20/index.tsx (hunk 5 above)
Comment (lalitkumarrock) [reply src/components/SendErc20/index.tsx]: Fixed it.

File: src/components/SendErc20/index.tsx (hunk 6 above)
Comment (lalitkumarrock) [reply src/components/SendErc20/index.tsx]: Fixed it.

File: src/utils/bigNumber.ts (hunk 8 above)
Comment (lalitkumarrock) [reply src/utils/bigNumber.ts]: Fixed it.

File: src/utils/validation.ts (hunk 9 above)
Comment (lalitkumarrock) [reply src/utils/validation.ts]: Fixed it.

File: src/utils/validation.ts (hunk 10 above)
Comment (lalitkumarrock) [reply src/utils/validation.ts]: Fixed it.

File: src/components/SendErc20/Erc20Input.tsx [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+  const onKeyDown = () => {
+    if (!disableSendOnEnter) handleEnterPress && handleEnterPress();
+  };
+
+  return (
+    <>
+      <input
+        type="text"
+        name="value"
+        id={`etherspot-ui-erc-20-input`}
```
Comment (IAmKio) [inline src/components/SendErc20/Erc20Input.tsx]: Remember this must be unique. Appending a random number here would be good enough.

**Review by IAmKio — CHANGES_REQUESTED** [review]
Just one thing missing - we're nearly there

File: src/components/SendErc20/Erc20Input.tsx (hunk 11 above)
Comment (lalitkumarrock) [reply src/components/SendErc20/Erc20Input.tsx]: FYI. Id change on every input value.

**Review by IAmKio — APPROVED** [review]
LGTM

**Pull request merged**